### PR TITLE
Realurl support for sites, languages and currencies

### PR DESCRIPTION
**Pull request by gilbertsoft**
Here a preview of a extended realurl support. Not finished now, as said just a preview. Before I have to check the other pending realurl changes...

### CODE REVIEW

**Comment by aimeos** [comment]
That doesn't work well because the Aimeos tables can be in another database. You have to create an Aimeos context object and use the Aimeos locale managers to retrieve this data.

File: Classes/Custom/Realurl.php [hunk 1]
```diff
@@ -141,6 +133,26 @@ public function addAutoConfig( $params, &$pObj )
 	}
 
 	/**
+	 * Adds no_cache to configuration if not already defined
+	 *
+	 * @param array $configuration
+	 * @return void
+	 */
+	protected function addNoCache(array &$configuration) {
+		if ( array_search( 'no_cache', $configuration ) !== false ) {
```
Comment (aimeos) [inline Classes/Custom/Realurl.php]: That won't work because `array_search()` doesn't search in multi-dimensional arrays

Comment (gilbertsoft) [inline Classes/Custom/Realurl.php]: I know, I have to merge from the boostrap_branch ;-)

**Comment by gilbertsoft** [comment]
I already thought about using the aimeos context for this, that's the reason I PR this preview. Are you interested including this changes because I don't need this for the moment but in the near future. So if you like to have it finished I don't wait for my own needs...

**Comment by aimeos** [comment]
It's nothing to add immediately, especially as the language selector is not usable in TYPO3 ("L" parameter values are no language codes and managed by RealUrl in the future) and therefore out of scope. The site and currency parameter parameter remains and both are strings where no mapping for RealUrl is needed at all.

**Comment by gilbertsoft** [comment]
I tought about a mapping for the language, but T3 uses a two letter code only. So I have to make a join query to static_info_tables, but only for T3 7.6. Before a complete different implementation only with static_info_tables and if this is not available I don't know. So I will continue the work, but without priority. If you have some ideas or thought on this please tell...

**Comment by gilbertsoft** [comment]
Have a look at the new version, now aimeos data is loaded with a context object and I added a new ext config option to activate the writting of the no_cache param (by default it is disabled). But can be enabled by default too, the check for an existing entry works fine now :-) Some tuning has to be done and I have to finish the tests with the locales. 
But today was horrible, I updated realurl this night but the new version doesn't run at all. The whole day I searched for errors, but at the end I discovered that the problems are in the new realurl and not in my changes :-( Never saw a such buggy stable bug fix release!

**Comment by aimeos** [comment]
We can confirm that: RealUrl 2.x is totally unusable

**Comment by gilbertsoft** [comment]
Since begin of november I worked with 2.1.5 and had no bigger problems so far. Yesterday I updated to 2.1.7 (a day before they released 2.1.6!) and then the troubles started! The only thing I see is special pages types are formatted wrong, here some google bot request. By the way, my first aimeos webshop is launched now :-D

> 66.249.66.198 - - [14/Feb/2017:22:58:20 +0100] "GET /mein-konto/d/358/Set_Body_L%C3%A4tzli_Nuggikette_Silikon_Stern/6/watch/add/358/ HTTP/1.1" 200 4369 "-" "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"
> 66.249.66.169 - - [14/Feb/2017:22:59:48 +0100] "GET /list/f/2/Nuggiketten_Holz/ HTTP/1.1" 200 8450 "-" "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"
> 66.249.66.169 - - [14/Feb/2017:22:59:53 +0100] "GET /nc/stock/?type=191351524&ai%5Bs_prodcode%5D%5B0%5D=nk-h-auto-stern&ai%5Bs_prodcode%5D%5B1%5D=nk-h-b%C3%A4r-auto&ai%5Bs_prodcode%5D%5B2%5D=nk-h-b%C3%A4r-herz&ai%5Bs_prodcode%5D%5B3%5D=nk-h-fuchs-stern&ai%5Bs_prodcode%5D%5B4%5D=nk-h-fussball-stern&ai%5Bs_prodcode%5D%5B5%5D=nk-h-maus-mit-masche-stern&ai%5Bs_prodcode%5D%5B6%5D=nk-h-schmetterling-stern&ai%5Bs_prodcode%5D%5B7%5D=nk-h-traktor-stern&ai%5Bs_prodcode%5D%5B8%5D=nk-h-waschb%C3%A4r-stern HTTP/1.1" 200 778 "http://www.zwergenfabrik.ch/list/f/2/Nuggiketten_Holz/" "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"
> 66.249.66.166 - - [14/Feb/2017:22:59:54 +0100] "GET /nc/counts/f/2/Nuggiketten_Holz//?type=191351524 HTTP/1.1" 200 949 "http://www.zwergenfabrik.ch/list/f/2/Nuggiketten_Holz/" "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"
>

**Comment by gilbertsoft** [comment]
Here the shop config:

TYPO3 7.6.15
Aimeos 2017.01 + a couple of my contributions and later changes of you
Aimeos Distro 16.10.0
Realurl 2.1.5
Bootstrap 7.0.5

After a week of testing I would say realy stable, works great so far!

**Comment by aimeos** [comment]
Congratulations! :-)

RealUrl 2.1.5 has some serious issues too and it's extremely slow. We recommend to use 1.13.x instead which is the only stable one with no problems.

**Comment by gilbertsoft** [comment]
Ok, have to try the change now first on the test system...

**Comment by aimeos** [comment]
Site and currencyid are already strings and don't need to be mapped. Maybe this would be enough:
```
$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['realurl']['_DEFAULT']['preVars'][] = array('GETvar' => 'site');
$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['realurl']['_DEFAULT']['preVars'][] = array('GETvar' => 'currencyid');
```

**Comment by aimeos** [comment]
We close this because site and currencyid can be added simply by adding the two lines above and language needs to be done in TYPO3/boostrap_package because of the necessary ID/code mapping. The rest like no_cache should go into the ext_localconf.php of your project specific TYPO3 extension too.

**Comment by gilbertsoft** [comment]
Ok, will try when we start with the localized version of the shop. I hope it's so easy as you wrote with our 4 letter language codes ;-)

**Comment by gilbertsoft** [comment]
Ok, will try when we start with the localized version of the shop. I hope it's so easy as you wrote with our 4 letter language codes ;-)

**Pull request closed without merging**